### PR TITLE
AVRO-2123: Java duration logical type

### PR DESCRIPTION
**Pull request by opwvhk**
## What is the purpose of the change

Add logical type and conversion for ["Duration"](https://avro.apache.org/docs/1.11.0/spec.html#Duration), in the context of [AVRO-2123](https://issues.apache.org/jira/browse/AVRO-2123)

This is an alternative implementation for PR #1263.

Core in this change is a new implementation of [`java.time.TemporalAmount`](https://docs.oracle.com/javase/8/docs/api/java/time/temporal/TemporalAmount.html), which combines functionality of both `java.time.Period` and `java.time.Duration`. This is needed to cleanly implement the logical type.

## Verifying this change

This change adds tests for all changed functionality:
- `org.apache.avro.util.TimePeriodTest`
- `org.apache.avro.TestLogicalType#durationExtendsFixed12`
- `org.apache.avro.generic.TestGenericLogicalTypes#readWriteDuration`
- `org.apache.avro.TestLogicalType#uuidExtendsString` (boyscout refactoring)

## Documentation

- Does this pull request introduce a new feature? (~yes~ / **no**)
- If yes, how is the feature documented? (**not applicable** / ~docs / JavaDocs / not documented~)

### CODE REVIEW

File: lang/java/avro/src/main/java/org/apache/avro/util/TimePeriod.java [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+ * supports fixed ranges up to about 68 years, {@code TimePeriod} can only
+ * handle about 49 days.
+ * </p>
+ *
+ * <p>
+ * Given the time span of what we do,
+ * </p>
+ * supporting a little less than 358 million years), and cannot support negative
+ * time periods.
+ * </p>
```
[the remark this answers is not in the document]
Comment (opwvhk) [reply lang/java/avro/src/main/java/org/apache/avro/util/TimePeriod.java:71]: You're right. This paragraph should be removed.

**Review by clesaec — APPROVED** [review]
LGTM

**Comment by GFriedrich** [comment]
Hello @opwvhk,
first of all thanks for adding the duration support for the Java implementation.
Unfortunately I found that it is incomplete. (for more details see AVRO-4159)
Side note: The declined PR of #1263 contained this piece of code, but it seems it was missed to keep that.

**Comment by opwvhk** [comment]
Thank you for pointing this out, @GFriedrich!

I've opened a new PR (#3416) to address this.

**Pull request merged**